fawltydeps/extract_declared_dependencies.py

```diff
@@ -149,7 +149,7 @@ def parse_value(value: str) -> Iterator[DeclaredDependency]:
         # TODO: try leveraging RequirementsFile.from_string once
         #       pip-requirements-parser updates.
         # See:  https://github.com/nexB/pip-requirements-parser/pull/17
-        with NamedTemporaryFile(mode="wt") as tmp:
+        with NamedTemporaryFile(mode="wt", delete=False) as tmp:
```

Comment:
I don't think we want to pass `delete=False` on non-Windows? Even on Windows, we should maybe try to find a solution that does not leave behind this temporary file?

I looked up [the docs](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) on `NamedTemporaryFile`, and it has a lot to say about how the `delete` flag (and the new `delete_on_close` flag added in v3.12) works, but none of it applied directly to us, I believe: We only close the file once (when exiting the `with`-block), and we certainly don't attempt to re-open it after closing (hence it does not matter if the file is deletes on close or deleted on context exit).

However, it also has this to say about deleting temporary files on Windows, which I believe might be related to our situation:

> On Windows, if `delete_on_close` is false, and the file is created in a directory for which the user lacks delete access, then the [`os.unlink()`](https://docs.python.org/3/library/os.html#os.unlink) call on exit of the context manager will fail with a [`PermissionError`](https://docs.python.org/3/library/exceptions.html#PermissionError). This cannot happen when `delete_on_close` is true because delete access is requested by the open, which fails immediately if the requested access is not granted.

AFAIU the problem we're trying to work around here is that even though we are allowed to _create_ the temporary file (and write into it), we somehow need Administrator access on Windows in order to delete it.

It would be nice to play around with Python 3.12 on Windows to see if the new `delete_on_close` flag can somehow solve the issue we're seeing here. Still, we wouldn't be able to use that here any time soon...

I noticed the `TODO` in the preceding comment, and tried to chase down the upstream issue to see if maybe we could sidestep the temporary file creation altogether. That rabbit hole led me to [this comment](https://github.com/nexB/pip-requirements-parser/pull/19#discussion_r1379279880), showing an alternative solution that together with `delete=False` also tries to manually remove up the temp file afterwards. I don't know if this would solve our problem. In any case that PR is still open upstream, so we'd have to replicate the same fix here rather that rely on the upstream solution...

Reply:
> It would be nice to play around with Python 3.12 on Windows to see if the new delete_on_close flag can somehow solve the issue we're seeing here. Still, we wouldn't be able to use that here any time soon...

Yes, that is why I wil leave this idea for now, as we need to support older versions of Python. It will be definitely a neat solution.

I replaced:
```python
        with NamedTemporaryFile(mode="wt", delete=False) as tmp:
            tmp.write(value)
            tmp.flush()
            for dep in parse_requirements_txt(Path(tmp.name)):
                yield replace(dep, source=source)
```
with:
```python
        # https://github.com/nexB/pip-requirements-parser/pull/19#discussion_r1379279880
        tf = NamedTemporaryFile('wt', delete=False,
                        # we prefer utf8 encoded strings, but ...
                        # - must not change newlines
                        # - must not  change encoding, fallback to system encoding for compatibility
                        newline='', encoding=None)
        tf.write(value)
        tf.close()
        try:
            for dep in parse_requirements_txt(Path(tf.name)):
                yield replace(dep, source=source)
        finally:
            unlink(Path(tf.name))
```
(see 0eb7945ba7fa1f09794561748fe99f49279bc9d3 )